src/main/java/org/jenkinsci/remoting/engine/JnlpProtocol4Handler.java

```diff
@@ -152,7 +152,7 @@ public Jnlp4ConnectionState createConnectionState(@Nonnull Socket socket,
                 .filter(new ConnectionHeadersFilterLayer(headers, handler))
                 .named(String.format("%s connection from %s", getName(), socket.getRemoteSocketAddress()))
                 .listener(handler)
-                .build(new ChannelApplicationLayer(threadPool, handler))
+                .build(new ChannelApplicationLayer(threadPool, handler, headers.get(JnlpConnectionState.COOKIE_KEY)))
```

Comment:
This part and the corresponding additions to ChannelApplicationLayer are part of attaching the cookie to the channel properties on the controller side.